rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+
+### Wildcard Support
+
+Specifying permissions individually for each database object can be tedious and challenging to maintain. Wildcards provide a succinct and stable alternative, especially in the face of evolving schemas. It is encouraged that database engines support wildcard functionality for targets. This implementation may involve enumerating database objects and applying permissions to those matching specific patterns. Additionally, if a new object is created during a user session, the corresponding wildcard-based permission will not be automatically applied until a permission synchronization occurs. Implementing an automated permission synchronization triggered by schema changes can enhance the user experience.
+
+Unlike targets, permissions typically originate from a fixed set. While the inclusion of wildcards, as seen in the `ALL` permission, can still be beneficial, doing it at Teleport-level may overlap with database-level functionality, potentially diminishing utility.
+
+### Permission Lifecycle
+
+The feature is based on the automated database user creation feature, simplifying the analysis of permission lifetime.
+
+We define "permission synchronization" as the process of applying a predetermined set of permissions to the user. These permissions are calculated based on their current roles in conjunction with the existing database schema.
+
+1. Initially, when the user does not exist, they possess no permissions except for default ones.
+2. Upon connection, we create a user in a manner specific to the database. This involves applying a fixed set of roles to the user, as specified in `spec.allow.db_roles`.
+3. Permission synchronization occurs after the user is created or reactivated.
+4. Additional permission synchronizations can be triggered as necessary, such as on new connections, detected schema changes (which may be detected using database-specific methods), or based on a scheduled timer.
+5. For consistency, we ensure that new connections receive an equivalent set of permissions. Notably, permissions for other protocols or unsupported versions do not affect what is applied to the database. Changes in the database schema can lead to previously "identical" permission sets diverging, for example, when a new table is added. The permissions are equivalent if they are stable under schema changes.
+
+6. Once the last connection for the user is terminated, they are either removed or deactivated.
+7. When a user is removed, all permissions are automatically revoked. In cases where full removal is not possible (example: a user is the owner of database objects in Postgres), a process is in place to ensure that all permissions are removed from the user, regardless of their source. This approach acts as a precaution against potential lingering permissions and simplifies the management of permissions.
```

Comment:
What is this process?

Reply:
Hmm, yeah this came out rather vague. This is implementation-specific but with current implementations we trigger a stored procedure that drops privileges and roles. I'll rephrase this.

Reply:
I know we are not able to prevent tampering with this stored procedure.  But can we do any detection to ensure it's running and validate it has not been tampered with?